### PR TITLE
Numpy batched fetch api

### PR DESCRIPTION
**Pull request by yaxxie**
This is an initial PR to get tests run and feedback against the implementation as it stands for batched numpy fetches, including a python call to do the same.

I rewrote `fetchallnumpy()` -- it looks the same to the user but under the hood is grabbing the generator and concatenating all the chunks together
I implemented `fetchbatchnumpy()` -- This immediately returns the next batch to be fetched
Both these methods are supported under the hood by the generator function `_numpy_batch_generator()` -- this returns a generator object which returns an array of masked arrays for each `next()`, without putting it into an ordered dict (this is left to the user facing methods to do)

I did not write any integration tests yet, as I would like some guidance on what you would like to see.

All feedback appreciated

### CODE REVIEW

**Comment by yaxxie** [comment]
Will fix the implementation so that the `test_numpy_empty_column` integration test does not break.

@MathMagique would you mind to let me know what else you would like to see here?

File: cpp/turbodbc_numpy/Library/src/numpy_result_set.cpp [hunk 1]
```diff
@@ -48,10 +48,15 @@ numpy_result_set::numpy_result_set(turbodbc::result_sets::result_set & base) :
 {
 }
 
-
-pybind11::object numpy_result_set::fetch_all()
+pybind11::object numpy_result_set::fetch_next_batch()
 {
 	std::size_t rows_in_batch = base_result_.fetch_next_batch();
+
+	if (!rows_in_batch)
```
Comment (MathMagique) [inline cpp/turbodbc_numpy/Library/src/numpy_result_set.cpp]: please use `not` instead of `!`, easier to read

File: python/turbodbc/cursor.py [hunk 2]
```diff
@@ -122,14 +123,34 @@ def fetchmany(self, size=None):
         return [row for row in islice(self, size)]
 
     def fetchallnumpy(self):
+        if _has_numpy_support():
```
Comment (MathMagique) [inline python/turbodbc/cursor.py]: Seems like a duplicate since the same check is part of the batch generator

**Comment by MathMagique** [comment]
The existing tests assert that the `fetchallnumpy()` functionality works as expected (and as before), so these tests should not need modification at all.

I would prefer the `fetchbatchnumpy()` functionality to return a generator. This would allow users to iterate over all batches in a natural way. The name should be adjusted to `fetchnumpybatches()` or similar, then. Still, it probably makes sense to keep the low-level generator object that does not perform the `OrderedDict` transformation.

I am missing some integration tests for the new batch functionality. I would like to see a test where you fetch a total of 5 rows (say integers `[1, 2, 3, 4, 5]`) in batches of two rows. Then check that the first batch contains `[1, 2]`, the next `[3, 4]`, and the last batch `[5]`. Also check there is no fourth batch. Checking a single data type should be sufficient, since all data types and null values are already checked by the previous `fetchallnumpy()` functionality (and thus tests the low-level implementation). Also, I'd like to see some tests with the new functionality for empty result sets.

For the tests, you can set the `rows_to_buffer` parameter to `2` when calling the `open_connection` or `open_cursor` context managers (https://github.com/blue-yonder/turbodbc/blob/master/python/turbodbc_test/helpers.py#L86). This should give you the desired level of control.

Also, I am missing some elementary documentation in the README file. An entry in the changelog would be great as well (probably version 1.1.1 will be the next one). Don't worry about the Appveyor build just now; this is work in progress.

File: cpp/turbodbc_numpy/Library/src/numpy_result_set.cpp (hunk 1 above)
Comment (yaxxie) [inline cpp/turbodbc_numpy/Library/src/numpy_result_set.cpp]: Sorry, I didn't realise this was a thing in c++11

Comment (MathMagique) [inline cpp/turbodbc_numpy/Library/src/numpy_result_set.cpp]: It is not a C++11 thing, it's probably just my thing. I also don't like `&&` (use `and`) or `||` (use `or`). I'm just glad that Python got this part right ;-).

**Comment by yaxxie** [comment]
Thanks for the feedback, will work on these and get back to you

File: cpp/turbodbc_numpy/Library/src/numpy_result_set.cpp (hunk 1 above)
Comment (keitherskine) [inline cpp/turbodbc_numpy/Library/src/numpy_result_set.cpp]: @MathMagique I like your thinking!  Clear code is key.  The fewer hieroglyphics the better.

**Comment by yaxxie** [comment]
I got a little problem

In numpy_result_set.cpp we have this:
```
	if (not rows_in_batch)
	{
		throw pybind11::stop_iteration();
	}
```

This is problematic because there is no empty "typed" list being returned any longer, via 

```
	std::vector<std::unique_ptr<masked_column>> columns;
	for (std::size_t i = 0; i != n_columns; ++i) {
		columns.push_back(make_masked_column(column_info[i].type));
	}
```

so when we retrieve the result set, we don't get anything except a StopIteration exception. The reason why I have a problem with this, is that this differs in the previous behavior of returning empty **typed** arrays, which we cannot do any longer.

Should we
1) Not raise the stop iteration, initialise the empty list, and catch the fact that its empty in the python and modify the generator?
2) Not be concerned about this?
3) Do something else?

File: cpp/turbodbc_numpy/Library/src/numpy_result_set.cpp (hunk 1 above)
Comment (yaxxie) [inline cpp/turbodbc_numpy/Library/src/numpy_result_set.cpp]: I don't disagree, this opened my eyes (I had no idea this was legal in C++)

**Comment by MathMagique** [comment]
Interesting. I'd opt for solution 1 you proposed. In the context of what was before, this makes sense. Also, I think the Python result set returns an empty list when no more results could be retrieved, so this would be consistent, at least.

[If no previous behavior had been given, your initial solution would have been acceptable behavior as well.]

**Comment by yaxxie** [comment]
So, c71bbb6 was to do option 1, and the unit tests pass. I feel like the code is a bit messy as a consequence but if you have no complaints I'll leave it. Working on the additional integration tests and documentation, but I'm not sure where the changelog is.

**Comment by MathMagique** [comment]
The changelog can be found in the root directory: `CHANGELOG.md`

**Comment by yaxxie** [comment]
Please review 84b6aab, and the changes as a whole. Provided the new integration tests pass, are you happy with this as it stands or would you like to see anything else?

**Comment by MathMagique** [comment]
@yaxxie I'll make the thorough review as soon as I can; probably not today.

**Comment by yaxxie** [comment]
@MathMagique were you able to take a look yet?

**Comment by MathMagique** [comment]
not yet, sorry :-(

File: python/turbodbc/cursor.py [hunk 3]
```diff
@@ -7,23 +7,24 @@
...
 
-
 def _has_numpy_support():
+
     try:
         import turbodbc_numpy_support
         return True
     except ImportError:
-        return False
+         return False
```
Comment (MathMagique) [inline python/turbodbc/cursor.py]: The indentation here is one space too much :-D

File: CHANGELOG.md [hunk 4]
```diff
@@ -21,6 +21,8 @@ Version 1.1.0 (unreleased)
     In particular, a few `long` types were converted to `intptr_t` and `int64_t`
     where appropriate. In particular, this affects the `field` type that may be used
     by C++ end users (so they exist).
+*   Added new cursor.fetchnumpybatches method which returns a generator to
```
Comment (MathMagique) [inline CHANGELOG.md]: please use backticks and brackets to mark this as a function: `cursor.fetchnumpybatches()`

File: README.md [hunk 5]
```diff
@@ -160,6 +160,19 @@ Here is how to retrieve a result set in the form of NumPy arrays:
                                     mask = [False False],
                                     fill_value = 1e+20))])
 
+You can also fetch numpy result sets in batches, based on the `read_buffer_size` attribute on the connection, using an iterable:
```
Comment (MathMagique) [inline README.md]: Please use the established way of writing NumPy in this document

File: python/turbodbc/cursor.py [hunk 6]
```diff
@@ -7,23 +7,24 @@
...
-def _make_masked_array(data, mask):
+def _make_masked_arrays(result_batch):
     from numpy.ma import MaskedArray
     from numpy import object_
-    if isinstance(data, list):
-        return MaskedArray(data=data, mask=mask, dtype=object_)
-    else:
-        return MaskedArray(data=data, mask=mask)
-
+    numpy_masked_array = []
```
Comment (MathMagique) [inline python/turbodbc/cursor.py]: please use the plural form `numpy_masked_arrays` (or even leave the `numpy` part out)

File: python/turbodbc/cursor.py [hunk 7]
```diff
@@ -122,17 +123,34 @@ def fetchmany(self, size=None):
...
-            return OrderedDict(columns)
         else:
             raise Error("turbodbc was compiled without numpy support. Please install "
                         "numpy and reinstall turbodbc")
+        numpy_result_set = make_numpy_result_set(self.impl.get_result_set())
+        firstrun = True
+        while True:
+            result_batch = numpy_result_set.fetch_next_batch()
+            batch_to_yield = _make_masked_arrays(result_batch)
+            next_batch = len(list(filter(len, batch_to_yield)))
```
Comment (MathMagique) [inline python/turbodbc/cursor.py]: I find the logic here a bit confusing. Basically, you want to know if this batch is empty or not. That could be done with a variable like this (unless I am missing something, obviously):
```
is_empty_batch = (len(batch[0]) == 0)
```

File: python/turbodbc/cursor.py [hunk 8]
```diff
@@ -122,17 +123,34 @@ def fetchmany(self, size=None):
...
-                          [_make_masked_array(data, mask) for data, mask in numpy_result_set.fetch_all()])
-            return OrderedDict(columns)
         else:
             raise Error("turbodbc was compiled without numpy support. Please install "
                         "numpy and reinstall turbodbc")
+        numpy_result_set = make_numpy_result_set(self.impl.get_result_set())
+        firstrun = True
+        while True:
+            result_batch = numpy_result_set.fetch_next_batch()
+            batch_to_yield = _make_masked_arrays(result_batch)
```
Comment (MathMagique) [inline python/turbodbc/cursor.py]: `batch_to_yield` is a bit of an unwieldy name, probably because the good name was already taken a line above. You could combine both lines into one:
```
batch = _make_masked_arrays(numpy_result_set.fetch_next_batch())
```

File: python/turbodbc_test/test_select_numpy.py [hunk 9]
```diff
@@ -48,9 +48,21 @@ def test_numpy_empty_column(dsn, configuration):
             cursor.execute("SELECT a FROM {}".format(table_name))
             results = cursor.fetchallnumpy()
             assert isinstance(results, OrderedDict)
-            assert len(results) == 1
+            assert len(results) == 1 # ncols
             assert isinstance(results[_fix_case(configuration, 'a')], MaskedArray)
 
+@for_each_database
+def test_numpy_empty_column_batch_fetch(dsn, configuration):
```
Comment (MathMagique) [inline python/turbodbc_test/test_select_numpy.py]: good test!

File: python/turbodbc_test/test_select_numpy.py [hunk 10]
```diff
@@ -109,6 +121,19 @@ def test_numpy_binary_column_larger_than_batch_size(dsn, configuration):
...
 
+@for_each_database
+def test_numpy_batch_fetch(dsn, configuration):
+    with open_cursor(configuration, rows_to_buffer=2) as cursor:
+        with query_fixture(cursor, configuration, 'INSERT INTEGER') as table_name:
+            cursor.executemany("INSERT INTO {} VALUES (?)".format(table_name),
+                               [[1], [2], [3], [4], [5]])
+            cursor.execute("SELECT a FROM {} ORDER BY a".format(table_name))
+            batches = cursor.fetchnumpybatches()
+            expected_list = [MaskedArray([1, 2], mask=False), MaskedArray([3, 4], mask=False), MaskedArray([5], mask=False)]
```
Comment (MathMagique) [inline python/turbodbc_test/test_select_numpy.py]: please use line breaks to separate the batches visually:

```
expected_list = [MaskedArray([1, 2], mask=False),
                 MaskedArray([3, 4], mask=False),
                 MaskedArray([5], mask=False)]
```
Also, `expected_list` should probably be renamed to `expected_batches`

File: python/turbodbc/cursor.py (hunk 7 above)
Comment (yaxxie) [inline python/turbodbc/cursor.py]: Very happy to change, but just to explain, hard coded indexes make me a little twitchy, so I sought a way to determine empty batched that did not use a hardcoded indexing, although I can't think of a time when a result set will not have even 1 column...

File: python/turbodbc_test/test_select_numpy.py (hunk 9 above)
Comment (yaxxie) [inline python/turbodbc_test/test_select_numpy.py]: Thanks!

File: python/turbodbc/cursor.py (hunk 7 above)
Comment (MathMagique) [inline python/turbodbc/cursor.py]: Yeah, indexes can be weird. Result sets are guaranteed to have at least one column, see this line of code: https://github.com/blue-yonder/turbodbc/blob/master/cpp/turbodbc/Library/src/command.cpp#L35

**Comment by MathMagique** [comment]
Thanks @yaxxie for this wonderful contribution! It actually added flexibility, new usability patterns, _and_ it simplified the code. On second thought, this is probably how I should have done it the first time around ;-). Great job!

**Comment by yaxxie** [comment]
Thanks for all the help, it was much appreciated!
